File: 112-path-sum/path-sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {boolean}
 */
var hasPathSum = function(root, targetSum) {
    if (!root) return false;

    const _dfs = (node, sum = 0) => {
        if (!node) return false;  
        sum += node.val;

        if (!node.left && !node.right) {
            if (sum === targetSum){
                return true;
            } else {
                return false;
            }
        }
        return _dfs(node.left, sum) || _dfs(node.right, sum);
    }
    return _dfs(root)
};

File: 112-path-sum/path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root: return False

        def _dfs(node, sum, target):
            if not node: return False
            sum += node.val

            if not node.left and not node.right:
                if sum == target:
                    return True
                else:
                    return False
            
            return _dfs(node.left, sum, target) or _dfs(node.right, sum, target)
        
        return _dfs(root, 0, targetSum)